docs/proposals/hpa/federated-hpa-v2.md

```diff
@@ -0,0 +1,809 @@
+  hpaAssignmentPolicyConfiguration:
+    targetCluster:
+      - clusterNames:
+          - member1
+        priority: 2
+      - clusterNames:
+          - member2
+        priority: 1
+#Cluster availableReplicas
+member1: 20
+member2: 1
+#ResourceBinding
+clusters:
+  - name: member1
+    replicas: 15
+```
+
+So after the assignment, the result of the HPA assignment will be:
+```
+#member 1
+minReplicas: 8
+maxReplicas: 20
+#member 2
+minReplicas: 1
+maxReplicas: 4
+```
+
+And if the sum of availableReplicas is smaller than FederatedHPAMaxReplicas, the left will be assigned to the cluster with the highest priority.
+
+The resource binding will be updated after the assignment, same with `StaticWeighted assignment policy`.
+
+
+### Autoscaling across clusters
+With this design architecture, the behavior of scaling across cluster containers has two parts: Scaling Up/Scaling down.
+And also, to make the FederatedHPA controller scaling work(no conflicting), FederatedHPA controller only can scale up the workload in member clusters.
+
+#### Duplicated assignment policy
+##### Scaling up
+FederatedHPA controller should list&watch the status of member HPA, when one member cluster could not scale up the workload(HPA status will show), the FederatedHPA controller will scale up the workload in other member clusters(the one with the most available replicas).
+
+The number of replicas that need to be scaled up in different clusters will be calculated from the member cluster's HPA status(metrics or pending pods exists)
+
+##### Scaling down
+FederatedHPA controller should not care about this.
+
+#### StaticWeighted assignment policy
+##### Scaling up
+Similar to `Duplicated assignment policy`, but FederatedHPA controller will scale up the replicas in member clusters based on the static weight.
+
+For example, there is the following cluster's status:
+```
+targetUtilization: 50%
+cluster 1 static weight: 1
+cluster 2 static weight: 3
+cluster 3 static weight: 1
+
+cluster 1:
+  current replicas: 4
+  maxReplicas: 4
+  current Utilization: 100
+
+cluster 2:
+  current replicas: 8
+  maxReplicas: 100
+  current Utilization: 50
+
+cluster 3:
+  current replicas: 8
+  maxReplicas: 100
+  current Utilization: 50
+
+```
+
+So, FederatedHPA controller will directly scale 3 replicas up in cluster 2, and 1 replica in cluster 3(If there are pending pods, the behavior will be the same).
+
+##### Scaling down
+FederatedHPA controller should not care about this.
+
+#### DynamicWeighted assignment policy
+##### Scaling up
+Similar to `StaticWeighted assignment policy`, but FederatedHPA controller will scale up the replicas in member clusters based on the dynamic weight(availableReplicas).
+
+For example, there is the following cluster's status:
+```
+targetUtilization: 50%
+
+cluster 1:
+  current replicas: 4
+  maxReplicas: 4
+  current Utilization: 100
+
+cluster 2:
+  current replicas: 8
+  maxReplicas: 100
+  current Utilization: 50
+  currentAvailableReplicas: 50
```

Comment:
How do we know the currentAvailableReplicas? Will FederatedHPA controller call estimator?

And how does FederatedHPA controller find out that member1 does not have available resource and need to be adjusted by FederatedHPA controller? By checking if there are any pending pods of this workload?

Reply:
> How do we know the currentAvailableReplicas? Will FederatedHPA controller call estimator?

In cluster CR, we can find the allocatable replicas. No, it will not call now, but I think we should in the future. `AvailableReplicas` is inaccurate, I describe it in the `Risks and Mitigations`.

> And how does FederatedHPA controller find out that member1 does not have available resource and need to be adjusted by FederatedHPA controller? By checking if there are any pending pods of this workload?

Yes, the pending workload triggers the operation, Or periodic.

Reply:
> In cluster CR, we can find the allocatable replicas.

Interesting, but this field is quite inaccurate. Workload could contain a node affinity based on which we could not determine `AvailableReplicas` by allocatable replicas.

> Yes, the pending workload triggers the operation, Or maxReplicas reached.

Could you enlighten me how to know if there is pending pods of this workload?

Reply:
> Interesting, but this field is quite inaccurate. Workload could contain a node affinity based on which we could not determine AvailableReplicas by allocatable replicas.

Yes, it's quite inaccurate, we need a better way to solve this. Do you have any good ideas?

> Could you enlighten me how to know if there is pending pods of this workload?

 the unavailable replicas in deployment status(long time pending), could get it from the control plane.

Reply:
What about talking to estimator and get the unschedulable pods of this workload?

Reply:
But estimator is an optional component. HPA should work ok by default.